### PR TITLE
Roc test can now support multiple files

### LINKED ISSUE #5103: Update `roc test` to support multiple files
[Zulip discussion](https://roc.zulipchat.com/#narrow/stream/231634-beginners/topic/Testing.20bulk.20Roc.20files/near/340057421)

All of the following seem like reasonable things to support:
- specifying multiple .roc files
- if you specify a folder, it's like running roc test on each .roc file in that folder
- allowing globs like *Blah.roc to specify multiple files

One important trick would be that when doing this we'd need to avoid duplicates - e.g. if multiple specified .roc files import the same (potentially unspecified) .roc file, we need to not run its tests more than once.

### PR DESCRIPTION
**Pull request by jfkonecn**
Closes #5103 

I didn't do any fancy. I just loops through all the passed file. Let me know if you want me to do something else.

### CODE REVIEW

**Comment by smores56** [comment]
Seems that you have a clippy error. Can you fix this and run `cargo clippy --workspace --tests -- --deny warnings` before committing to validate you've fixed the issue?

```
     Checking roc_cli v0.0.1 (/Users/m1ci/actions-runner2/_work/roc/roc/crates/cli)
error: this `else { if .. }` block can be collapsed
   --> crates/cli/src/lib.rs:641:16
    |
641 |           } else {
    |  ________________^
642 | |             if matches.get_flag(FLAG_VERBOSE) {
643 | |                 println!("Compiled in {} ms.", compilation_duration.as_millis());
644 | |                 for module_test_results in results_by_module {
...   |
651 | |             }
652 | |         }
    | |_________^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#collapsible_else_if
    = note: `-D clippy::collapsible-else-if` implied by `-D warnings`
    = help: to override `-D warnings` add `#[allow(clippy::collapsible_else_if)]`
```

Also, please sign your commits, a guide for which can be found in our [CONTRIBUTING.md document](https://github.com/roc-lang/roc/blob/main/CONTRIBUTING.md#commit-signing).

**Comment by jfkonecn** [comment]
I need to read the doc! :laughing:

**Comment by jfkonecn** [comment]
Debugging time!

**Comment by jfkonecn** [comment]
I'm not quite sure what was wrong with the pipeline. The test passing on my machine. I just did a rebase with the latest commit. I'm hoping that fixes it.

File: crates/cli/src/lib.rs [hunk 1]
```diff
@@ -498,18 +499,20 @@ pub fn test(matches: &ArgMatches, target: Target) -> io::Result<i32> {
         Some(n) => Threading::AtMost(*n),
     };
 
-    let path = matches.get_one::<PathBuf>(ROC_FILE).unwrap();
+    // let path = matches.get_one::<PathBuf>(ROC_FILE).unwrap();
```
Comment (lukewilliamboswell) [inline crates/cli/src/lib.rs:502]: Was this left here by mistake?

**Review by lukewilliamboswell — COMMENTED** [review]
Thank you for this @jfkonecn, it's looking good. I played with this a bit and found a few things that were unexpected. Would you mind looking into these?

I tried to run this on a folder but got an error. I was expecting this behaviour

> if you specify a folder, it's like running roc test on each .roc file in that folder

```
$ cargo run -- test examples/
── FILE PROBLEM in examples/ ───────────────────────────────────────────────────

I tried to read this file:

    examples/

But ran into:

    is a directory
```


I ran agains multiple files, it prints out `No expectations were found.` multiple times. I was expecting this to just print once.

```
$ cargo run -- test examples/platform-switching/*.roc
Running `target/debug/roc test examples/platform-switching/main.roc examples/platform-switching/rocLovesC.roc examples/platform-switching/rocLovesRust.roc examples/platform-switching/rocLovesWebAssembly.roc examples/platform-switching/rocLovesZig.roc`
No expectations were found.
No expectations were found.
No expectations were found.
No expectations were found.
No expectations were found.
No expectations were found.
```

When I specific only two files, it printed out three times. 

```
$ cargo run -- test examples/platform-switching/rocLovesZig.roc examples/platform-switching/rocLovesRust.roc
    Finished dev [unoptimized + debuginfo] target(s) in 0.41s
     Running `target/debug/roc test examples/platform-switching/rocLovesZig.roc examples/platform-switching/rocLovesRust.roc`
No expectations were found.
No expectations were found.
No expectations were found.
```

**Comment by jfkonecn** [comment]
> Thank you for this @jfkonecn, it's looking good. I played with this a bit and found a few things that were unexpected. Would you mind looking into these?
> 
> 
> 
> I tried to run this on a folder but got an error. I was expecting this behaviour
> 
> 
> 
> > if you specify a folder, it's like running roc test on each .roc file in that folder
> 
> 
> 
> ```
> 
> $ cargo run -- test examples/
> 
> ── FILE PROBLEM in examples/ ───────────────────────────────────────────────────
> 
> 
> 
> I tried to read this file:
> 
> 
> 
>     examples/
> 
> 
> 
> But ran into:
> 
> 
> 
>     is a directory
> 
> ```
> 
> 
> 
> 
> 
> I ran agains multiple files, it prints out `No expectations were found.` multiple times. I was expecting this to just print once.
> 
> 
> 
> ```
> 
> $ cargo run -- test examples/platform-switching/*.roc
> 
> Running `target/debug/roc test examples/platform-switching/main.roc examples/platform-switching/rocLovesC.roc examples/platform-switching/rocLovesRust.roc examples/platform-switching/rocLovesWebAssembly.roc examples/platform-switching/rocLovesZig.roc`
> 
> No expectations were found.
> 
> No expectations were found.
> 
> No expectations were found.
> 
> No expectations were found.
> 
> No expectations were found.
> 
> No expectations were found.
> 
> ```
> 
> 
> 
> When I specific only two files, it printed out three times. 
> 
> 
> 
> ```
> 
> $ cargo run -- test examples/platform-switching/rocLovesZig.roc examples/platform-switching/rocLovesRust.roc
> 
>     Finished dev [unoptimized + debuginfo] target(s) in 0.41s
> 
>      Running `target/debug/roc test examples/platform-switching/rocLovesZig.roc examples/platform-switching/rocLovesRust.roc`
> 
> No expectations were found.
> 
> No expectations were found.
> 
> No expectations were found.
> 
> ```

Thanks for the review! I'll take a look.

**Comment by jfkonecn** [comment]
@lukewilliamboswell I've addressed you comments.

I'm happy to do any code clean up you want me to do even if you think it's a minor issue.

**Comment by lukewilliamboswell** [comment]
Thank you @jfkonecn 

When I run it against the `examples/` directory it panics with the following. 
 
```
$ nix develop
$ cargo run -- test examples/
thread '<unnamed>' panicked at crates/compiler/solve/src/to_var.rs:507:17:
assertion failed: !tags.is_empty() || !ext_slice.is_empty()
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Comment by jfkonecn** [comment]
Alright more debugging! 😆

**Comment by lukewilliamboswell** [comment]
@agu-z -- would you be able to look at this? I'm interested to know if we should log a separate issue and merge this PR as is?

Below is the assertion that is being hit. This seems unrelated to the changes in this PR.

```
// An empty tags is inefficient (but would be correct)
// If hit, try to turn the value into an EmptyTagUnion in canonicalization
debug_assert!(!tags.is_empty() || !ext_slice.is_empty());
```

**Comment by jfkonecn** [comment]
@lukewilliamboswell on the latest commit of main does your code generate the same error if you run `cargo run -- test examples/ruby-interop/platform/main.roc`? That seems to be the issue for me.

**Comment by agu-z** [comment]
@lukewilliamboswell I'm sorry I hadn't seen this. Do you suspect this is related to one my changes?

**Comment by agu-z** [comment]
@lukewilliamboswell @jfkonecn You can reproduce it this on `main` by running `roc test` against the `examples/virtual-dom-wip` platform file:

```shell
roc test examples/virtual-dom-wip/platform/server-side.roc
```

When you run `roc test` against `examples/` this file is included so you run into that issue. I think we should unblock this PR and test only against the files in `examples/` that don't already fail on `main`.

There seems to be a `mono` test that reproduces this issue already (https://github.com/roc-lang/roc/issues/6947), so I don't think we need to create a separate one.

**Review by lukewilliamboswell — APPROVED** [review]

**Pull request merged**